s_tui/sources/rapl_read.py

```diff
@@ -32,27 +33,84 @@
-    ret = list()
-    for path in basenames:
-        name = None
-        try:
-            name = cat(pjoin(path, 'name'), fallback=None, binary=False)
-        except (IOError, OSError, ValueError) as err:
-            logging.warning("ignoring %r for file %r",
-                            (err, path), RuntimeWarning)
-            continue
-        if name:
+class RaplReader:
+    def __init__(self):
+        basenames = glob.glob('/sys/class/powercap/intel-rapl:*/')
+        self.basenames = sorted(set({x for x in basenames}))
+
+    def read_power(self):
+        """ Read power stats and return dictionary"""
+
+        pjoin = os.path.join
+        ret = list()
+        for path in self.basenames:
+            name = None
             try:
-                current = cat(pjoin(path, 'energy_uj'))
-                max_reading = 0.0
-                ret.append(RaplStats(name, float(current), max_reading))
+                name = cat(pjoin(path, 'name'), fallback=None, binary=False)
             except (IOError, OSError, ValueError) as err:
                 logging.warning("ignoring %r for file %r",
                                 (err, path), RuntimeWarning)
-    return ret
+                continue
+            if name:
+                try:
+                    current = cat(pjoin(path, 'energy_uj'))
+                    max_reading = 0.0
+                    ret.append(RaplStats(name, float(current), max_reading))
+                except (IOError, OSError, ValueError) as err:
+                    logging.warning("ignoring %r for file %r",
+                                    (err, path), RuntimeWarning)
+        return ret
+
+    @staticmethod
+    def available():
+        return os.path.exists("/sys/class/powercap/intel-rapl")
+
+
+class AMDEnergyReader:
+    def __init__(self):
+        inputs = list(zip((cat(filename, binary=False) for filename in sorted(
+                                           glob.glob('/sys/devices/platform/amd_energy.0/hwmon/hwmon*/energy*_label'))),
+                           sorted(glob.glob('/sys/devices/platform/amd_energy.0/hwmon/hwmon*/energy*_input'))))
+
+        # How many socket does the system have?
+        socket_number = sum(1 for label, _ in inputs if 'socket' in label)
+        inputs.sort(key=lambda x: self.get_input_position(x[0], socket_number))
+
+        self.inputs = [(self.get_pretty_label(label), inp) for label, inp in inputs]
+        #for i in range(len(self.inputs)):
+        #    self.inputs[i] = (self.get_pretty_label(self.inputs[i][0]), self.inputs[i][1])
+
+    @staticmethod
+    def match_label(label):
+        return re.search(r'E(core|socket)([0-9]+)', label)
+
+    @staticmethod
+    def get_input_position(label, socket_number):
+        num = int(AMDEnergyReader.match_label(label).group(2))
+        if 'socket' in label:
+            return num
+        else:
+            return num + socket_number
+
+    @staticmethod
+    def get_pretty_label(label):
```

Comment:
I am a bit concerned about search and replacing of labels. Mainly because it might be slightly different of each machine / kernel / distro.
Could you please tell me what do the labels look like if you are not manipulating them?

This is not necessarily a bad idea, but if they are readable as they are, it might be simpler to not alter them

Reply:
![grafik](https://user-images.githubusercontent.com/11158353/89764231-4f910180-daf4-11ea-876d-acfcf5c72a37.png)
This is what it looks like when the labels are used without further processing.